Task1

sergeyratz opened the pull request:
Task 1 implementation (pull requests)

Task 1
Распечатать произвольное число в двоичной системе исчисления.
Описание
Написать функцию, принимающую на вход произвольное целочисленное значение и возвращающую строку, содержащую его представление в двоичной системе исчисления.
Декларация
voidNumberAsBinary( char* _result,// строка с результатом unsignedint _number// произвольное число )
Ожидаемый результат
Например, для числа 13 данная функция должна вернуть строку "1101". Проверка работы. Решение должно демонстрировать работу функции на примере 10 случайных чисел в диапазоне 0 до 512.

Commit:
Answer for Task1

Commit:
Answer for Task1

Commit:
Answer for Task1

sergeyratz commented:
По-моему задание подразумевала написание функции с заданной сигнатурой:
`voidNumberAsBinary( char* _result,/* строка с результатом */ unsignedint _number/** произвольное число */)`

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
 
-int main(int argc, char** argv) {
+char result[11];
+int i = 0;
```
Декларирование глобальной **не константной** переменной, приводит к многим сайдефектам. В часности не очевидному сиквенсу. _(Модифицироваться переменная может в нескольких функциях, и порядок их вызова можешь знать только ты.)_ **Так лучше не делать.** Кроме того, это имя распостраненное для циклов, следовательно использовать i в данном скопе ты уже не сможешь.

sergeyratz replied:
Вообще здесь этого вообще не должно быть. Ну и это не глобальная штяжка.

a-nechyporenko replied:
как я понял "i" это размер твоей строки фактической строки. Измени название  этой переменной

KirillHelm replied:
Не, это глобальная, main начинается ниже.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+char result[11];
+int i = 0;
+
+void write_result()
```
название функций луше писать начиная с строчной и каждое слово с прописной. если это абривиатура, то прописными. нижнее подчеркивание допустимо как часто повторяющийся префикс типа request_getSomething.
В твоем случаи лучше назвать функцию **writeResult**

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+			cout << result[i];
+		}
+	
+		cout << "\n \n";
```
Вместа символа переноса строки (к слову пробел между ними не обязателен) можно использовать стандартный модификатор **endl**. То есть cout << "\n\n"; еквивалентен cout << endl << endl;

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+{
+		i = i-1;
+		cout << "This namber in Bin = ";
+		for(i; i >= 0; i--)
```
Всегда в циклах используй --i или ++i.
_(Почитай про префиксный и постфиксный инкремент и декремент, порядок выполнения действий в них https://lurkmore.to/++i_+_++i)_

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+
+void write_result()
+{
+		i = i-1;
```
вместо этого быстрее отработает и лучше написать --i;

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+	int changes;
+	int oneOrZero;
+	
+	for(int j = 0; j < 10; j++)
```
вот тут как раз и приходится использовать j вместо i, хотя можно было бы решить эту проблему если бы i не была в глобальном скопе.
Так же 10 - magic number, почему 10? откуда оно взялось? почему бы не 20? вопросы риторические, лучше создать константную переменную с понятным названием которая опишет почему именно столько итераций должно произойти и присвоить ей число 10. Так будет понятнее:
const unsigned int loopCount = 10;
for (int j = 0; j < loopCount; ++j)

либо можно её вшить сразу в скоуп только цикла типа:
for (int j = 0, loopCount = 10; i < loopCount; ++j)


### Also, По заданию расчётыдолжны были выполнятся в отдельной функции **numberAsBinary**, а не в цикле находящимся в main.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+			do
+			{
+				changes = giveDecNam / 2;
+					oneOrZero = giveDecNam - changes*2;
```
не забываем про табуляцию, что бы строки стояли на 1 уровне. _табуляцию лучше заменить 3 пробелами._

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+			}
+		while(giveDecNam != 0);
+		
+		write_result();	
```
что бы избежать как раз этой проблемы со скоупом, лучше записывать все число в строку, после чего выводить целую строку, вместо того, что бы формировать массив чаров **статический** и выводить по одному. **А что если у тебя будет число больше длиной чем 11 символов?** Скажут модифицировать пример и выводить числа до 1'000'000 в двоичной? Придется переписывать размер массива? А память на него то в стеке выделяется, зачем лишнию использовать. лучше уже сделать функцию типа _writeResult( char* text );_ что бы её не передал, она все выведет целиком. после чего спокойно можешь очистить память из под старой строки.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+					result[i] = '1';
+				}
+				giveDecNam = changes;
+				i++;
```
снова ++i

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+				giveDecNam = changes;
+				i++;
+			}
+		while(giveDecNam != 0);
```
выше ты писал правильно, **слева** была **константа**. лучше всегда писать константу слева, что бы избежать присваивания, если не пропечатаешь !.
То есть 0 != giveDecNam

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+	{
+		giveDecNam = 1 + rand() % 512;
+		cout << "Dec namber = " << giveDecNam << endl;
+			do
```
почему используешь do while, а не while? по возможности лучше использовать while, он более очевидный и понятный. _do while можно использовать в паттерне "do while (false)"_

NikolayPV replied:
Вначале стояло условие проверки while  (...>1)
И для числа 1 было необходимо do. Со временем код изменился.

a-nechyporenko replied:
Вместо всего этого обычно лучше использовать for.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,54 @@
+	int giveDecNam;
+	int showBin;
+	int changes;
+	int oneOrZero;
```
если переменная может принимать значение true или false (0/1) лучше использовать bool обыкновенный

Что до остальных переменных, у них очень неочевидные названия. лично мне посмотрев на них **не понятно** зачем они нужны, без копания в коде. Название переменных должно быть однозначное и понятное.

_К примеру giveDecNam лучше бы называлось randomeNumber и присутствовало только в где ты сетаешь на неё рандомное число и где её используешь в расчётах._

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,53 @@
+	{
+		giveDecNam = 1 + rand() % 512;
+		cout << "Dec namber = " << giveDecNam << endl;
+			do
```
Попробуй вынести это в функцию.

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,53 @@
+			}
+		while(giveDecNam != 0);
+		
+		write_result();	
```
добавь в свою функцию параметры(строку, ее размер) необходимые для вывода строки, это поможет тебе избавится от глобальных переменых.

Commit:
global changes in function Task1

- Function numberAsBinary is created
-The global variables were deleted.
-tab = 3 spaces.
-added the counter of Bits for Dec nambers

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+//-----------------------------------------------------------------------------------------
+int main() 
+{
+	const int MAX_SIZE_BIN_NAMB = 10;
```
MAX_SIZE_BIN_COUNT если на то пошло. Но название все равно не понятное
Максимальный размер бинарных чисел/кол-ва. Что ты этим хотел сказать?
Если максимальную кол-во бинарных чисел то:
MAX_BIN_COUNT
Если максимальную длину бинарного числа то:
MAX_BIN_NUMBER_LENGTH
В нейменге мешанина и не понятно за что он отвечает
К слову можно инициализировать как:
const int MAX_SIZE_BIN_COUNT(10);

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
 
-int main(int argc, char** argv) {
+//-----------------------------------------------------------------------------------------
+void numberAsBinary( unsigned int number, char *_result, int looper);
```
не понятные названия переменных, что значет loper? зачем нижний прочерк только у result?
инты передавай по const &

NikolayPV replied:
инты передавай по const &
А зачем? У меня нет цели менять, я передаю по значению т.к. нужно только само число и не собираюсь менять то, что в main.  А вот в numberAsBinary буду менять.

KyryloRud replied:
Это хороший тон, читай внимательно комменты, писал уже что:
1. Передавать нужно все что передается передавать по ссылке. Исключения составляют только указатели и переменные типа bool (но тут на твой вкус). **Благодаря ссылке у тебя не создается локальная копия объекта для функции** __для примера, допустим тебе нужно обработать объект который занимает в памяти ~10мб, передав по значению ты получишь его автоматически созданный инстанс и в памяти уже вместо 1 объекта будет 2 и памяти уже затратится 20мб.__
2. Константа нужна что бы **случайно не модифицировать объект** переданный по ссылке, таким способом ты решаешь 2 проблемы, излишнюю растрату памяти, и вытекающую из неё - случайную модификацию.

**Читаем Маерса** (на данный момент не важно какое издание, это актуально везде):
Константы везде где только можно: https://www.e-reading.club/chapter.php/1002058/15/Mayers_-_Effektivnoe_ispolzovanie_CPP.html
Передача по константной ссылке: https://www.e-reading.club/chapter.php/1002058/54/Mayers_-_Effektivnoe_ispolzovanie_CPP.html

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+#include <cstring>
 
-/* run this program using the console pauser or add your own getch, system("pause") or input loop */
+using namespace std;
```
просто как возможный кейс, менять не нужно, но для справки. Не всегда стоит подключать все пространство имен, если к примеру ты пользуешь из std только поток вывода cout, то можно вместо using namespace std; написать using std::cout; using std::endl; (если ты пользуешь перенос строк)

a-nechyporenko replied:
Ну или можно указывать для каждой функции ее наймспейс(так более наглядно какая именно функция используеться а так же где она находиться).

KyryloRud replied:
это в случаи если **namespace** - кастомный __(чаще всего, не всегда)__, согласен, это полезно в некоторых случаях

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+{
+	const int MAX_SIZE_BIN_NAMB = 10;
+	char result[MAX_SIZE_BIN_NAMB];
+	unsigned int randDecNamb;
```
нейминг можно не сокращать так, желательно вкладываться в 80 символов в строке, но тут спокойно можно намисать randomDecimalNumber; но тогда уж лучше Digit а не Number - но это уже на твой выбор

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+		
+		cout << "This namber in Bin = " << result << endl << endl;
+		
+		memset(result, 0, MAX_SIZE_BIN_NAMB*sizeof(int)); 
```
лучше вынести MAX_SIZE_BIN_NAMB*sizeof(int) в отдельную переменную, не очевидно что это ты туда передаешь.
То есть сейчас ни из названия переменной в сигнатуре функции (looper), ни из того что ты туда передал (MAX_SIZE_BIN_NAMB*sizeof(int)) мне не понятно что это и для чего это, и нужно лезть в код функции что бы понять как её нужно юзать. Из её сигнатуры должно быть однозначно ясно для чего она и что она возращает (из названия и return value type) и что ей нужно подать на вход (название параметров и их типы)

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+		 
+		while(looper >= 0)
+		{	
+			int changes = number / 2;
```
changes дальше не модифицируется, его можно и нужно сделать константой

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,56 @@
+		while(looper >= 0)
+		{	
+			int changes = number / 2;
+			int oneOrZero = number - changes*2;
```
название тоже лучше поменять на bit или что-то типа того. тогда будет ясно что это такое, а то пока что это какая-то неопрделенность. Так же не модифицируешь -> делаешь константой

Commit:
fixed  bugs in Task 1
- the announcement of variables outside of cycles
- the variable looper in function numberAsBinary was renamed
- constant MAX_SIZE_BIN_NAMB in function main was renamed
- aded the new constant in function main.
- added the obvious transformation of variables type.
- changed while to for
- changed the type of work with array

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,56 +1,67 @@
 		cout << "This namber in Bin = " << result << endl << endl;
 		
-		memset(result, 0, MAX_SIZE_BIN_NAMB*sizeof(int)); 
+		cleanMass = MAX_BIN_NUMBER_LENGTH*sizeof(int);
```
Зачем считать каждый раз одно и тоже число?

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,56 +1,67 @@
 	{
-		randDecNamb = 1 + rand() % 512;
-		counterBitInDec = log2(randDecNamb);
+		randomDecimalNamber = 1 + rand() % 512;
```
Мне кажеться можно вынести 27 и 29 строку в отдельную функцию, которая будет генерить рандомное число

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,56 +1,67 @@
-	char result[MAX_SIZE_BIN_NAMB];
-	unsigned int randDecNamb;
+	const int MAX_BIN_NUMBER_LENGTH = 10;
+	const int QUANTITY_DEC_DIGITS = 10;
```
я так понял что это количество чисел, которые ты оброботаешь и вывидешь, если это так то название мне кажеться можно поменять на более читаемое

Commit:
- changed he program algorithm
- removed excess variable

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,60 @@
+	const int MAX_BIN_NUMBER_LENGTH = 10;
+	const int QUANTITY_DEC_DIGITS = 10;
+	unsigned int randomDecimalNamber;
+	int cleaningArray;
```
cleaningArray - писал в прошлый раз, уже намного лучше, но название этой переменной не то. Это **verb** и из него не понятно смысл переменной. Она должна быть **noun**.

KyryloRud replied:
Продолжая тему **memset**. Вот линка на доку: http://www.cplusplus.com/reference/cstring/memset/?kw=memset
Прошу заметить что там третий параметр называется **num** = **number** никак не глагол.
Так что передавая переменную с названием **cleaningArray** _(процесс очистки массива, даже функция будет называться cleanArray, а это скорее состояние)_ не понятно что ты подразумеваешь.

> `void * memset ( void * ptr, int value, size_t num );`
> Sets the first _num_ bytes of the block of memory pointed by _ptr_ to the specified _value_ (interpreted as an unsigned char).

Из доки ясно, что она сетает кол-во байт, следовательно это будет что-то связанное с **count** уже, а не с **cleaning**.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,60 @@
+	unsigned int randomDecimalNamber;
+	int cleaningArray;
+	
+	char *FindFirstUsed1;
```
Советуют писать указатель вместе с типом, а не названием переменной (не могу поручится, Маерс это говорил ли или кто другой). Но писать лучше **char\* var**, а не **char \*var** по тому, что тип у тебя **char***, а не просто **char**.

Кроме того, цифры лучше не использовать в названиях переменных, их легко не заметить, для наглядности:
**variable1
variable2**
Не так читается по сравнению
**firstVariable
secondVariable**

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,60 @@
+		FindFirstUsed1 = strchr(result, '1');
+		cout << "This number in Bin = " << FindFirstUsed1 << endl;
+
+		cleaningArray = MAX_BIN_NUMBER_LENGTH*sizeof(int);
```
Можешь объяснить для чего ты умнажаешь длину числа на 4?

NikolayPV replied:
4 в смысле int? 
Если да, то для определения полной длины, а после очистки массива.
Как она работает до конца не понимаю. ее цель - очистить массив после окончания работы.

KyryloRud replied:
# Смотри что ты делаешь:

1. Строка 34. Тут у тебя константа `MAX_BIN_NUMBER_LENGTH`, которая равна 10, умнажается на размер integer в текущей системной конфигурации. Сейчас все это стандартизированно и размер **int** будет 4 байта.
Следовательно переменной `cleaningArray` ты присваиваешь значение 10 * 4 = 40 byte.
`cleaningArray = MAX_BIN_NUMBER_LENGTH*sizeof(int);`

2. Строка 35: Ты передаешь значение cleaningArray в функцию memset [(ссылка на документацию)](http://www.cplusplus.com/reference/cstring/memset/?kw=memset), вот что она говорит:
> void * memset ( void * ptr, int value, size_t num );
> Fill block of memory
> Sets the first num bytes of the block of memory pointed by ptr to the specified value (interpreted as an unsigned char).

Устанавливает в первые **num** байт в блок памяти по указателю **ptr** переданное значение **value**.
Из этого следует, что ты пытаешься ей _(функции)_ передать блок памяти по адресу `result`, сетнуть туда значение `0` в его _(в смысле адресса)_ первые `cleaningArray = 40` _(как мы выяснили из первого пункта)_ байт.
`memset(result, 0, cleaningArray);`

# Что нужно сделать вместо этого:
Избавиться от переменной `cleaningArray`, как минимум в аргументе функции и передавать ей константу `MAX_BIN_NUMBER_LENGTH`. По тому, что у тебя размер под массив `result` выделен такой же (см. строка 22). А ты пытаешься заполнить нулями размер массива в 4 раза больший. 
`char result[MAX_BIN_NUMBER_LENGTH];`

# WARNING!
Как показала практика, при выводе функция размер массива не увеличет, НО! **Функция начнет сетать байты за пределами массива и в случаи если там находится сейчас информация о другом объекте, то ты побьешь память, а это приведет либо к ambiguous behavior либо к segmentation fault.**

KyryloRud replied:
К слову, у тебя это не очистка массива, так что называть переменную размером равную 4м массивов - некорректно `cleaningArray`. Исправь это, а точнее как указал выше в комменте - удали вообще её.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,60 @@
+	int mask = 512;
+	
+	cout << "Dec number = " << number << endl;
+	for ( int i = 0; i <= 9; ++i )
```
Тут ошибка которой ты не допустил в цикле **main** частично:
1. Снова история с **Magic Number**, девятки там быть не должно.
2. **Константа** всегда **слева**.
```
const int bitsCount = 10;   //я использовал 10, если 9 более информативно можно и девяточку
for (int i = 0; bitCount > i; ++i)   //тогда естественно будет bitCount >= i
```

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,60 @@
+	cout << "Dec number = " << number << endl;
+	for ( int i = 0; i <= 9; ++i )
+	{
+		if ((number & mask) > 0)
```
Старайся ствить пробельчики когда накапливается условие со вложенными скобками, вот такое читаться будет лучше:
`if ( (number & mask) > 0 )`

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -3,64 +3,57 @@
+			result[i] = '0';
 		}
-		number = changes;
+		mask = (mask >> 1);
```
Вопервых не понятно длячего тут скобки? Оператор шифта вправо по преоритету выше нежели оператор присваивания, так что это смысла не имеет. Кроме того, ты делаешь шифт на 1 и больше у тебя нет никаких операций, так что проще и читаемее будет записать это как:
mask >>= 1;

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -3,64 +3,57 @@
-	cout << "Dec namber = " << number << endl;
-	int changes;
-	int oneOrZero;
+	int mask = 512;
```
mask -> bitMask тоже нейминг

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -3,64 +3,57 @@
 
 //-----------------------------------------------------------------------------------------
-void numberAsBinary(unsigned int number, char *result, int indexLocationInMass)
+void numberAsBinary(unsigned int number, char *result)
```
тут тоже самое с **char\*** а не **char \***

Commit:
code optimization

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+	unsigned int randomDecimalNamber;
+
+	char* result;
+	result = new char[MAX_BIN_NUMBER_LENGTH];
```
Где ты очищаешь эту память? Ну и чисто мое мнение, как по мне логичнее что бы память выделяла сама функция

KyryloRud replied:
+1 Но не забывай, передавая указатель в функцию, и выделяя память заново, можешь потерять адрес памяти уже под выделенную. По этому нужно быть аккуратным с работой с таким вариантом действий.

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+
+	for(int j = 0; j < QUANTITY_DEC_DIGITS; ++j)
+	{
+		cin >> randomDecimalNamber;
```
что будет если пользователь введет невалидное значение?

KyryloRud replied:
кроме невалидного значения (просто раскрываю тему) не забываем про то, что пользователь может ввести буквы.
создай функцию типа `bool isStringCorrespondUnsignedIntValue(char* string)` и если можно корректно конвертнуть строку в `unsigned int` -> `return bool = true;` ну и соответственно, если строка содержит любые символы кроме цифр и невозможно её уместить в `unsigned int` -> `return bool = false;`. И на основе этого значения хендли дальнейшую работу.

NikolayPV replied:
выведется прошлый ответ

KyryloRud replied:
Это ещё почему? Объясни пожалуйста по какой причине должно выводится предыдущий ответ?

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
 }
+
+//-----------------------------------------------------------------------------------------
+void numberAsBinary(unsigned int number, char* result)
```
думаю в названии функции можно указать деапозон валидных чисел и хорошо бы сделать инт константным, если ты его не меняешь

a-nechyporenko commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+		
+	if(0 != result && number <= bitMask)
+	{
+		while (bitMask > 0)
```
возможно если вместо while использовать for то код был бы более прозрачным, и можно было бы нормально итерировать масив

KyryloRud replied:
тут лично я не согласен, я полагаю что `while` - вполне себе ещё могёт и тут в самый раз :))

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
 //-----------------------------------------------------------------------------------------
 int main() 
 {
 	const int MAX_BIN_NUMBER_LENGTH = 10;
```
Тебе хватит 1й переменной. Зачем две одинаковые? Причем по одной ты выделяешь память, другую же используешь в цикле. **Оставляй одну.**

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
 
 //-----------------------------------------------------------------------------------------
-void numberAsBinary(unsigned int number, char *result)
+void numberAsBinary(unsigned int number, char* result)
```
Уже писал не раз, само значение `unsigned int number` ты менять не будешь, да и к тому же почему передаешь по значению?
`const unsigned int&`
_(проверяй пожалуйста перед тем как комитить, уже не 1 раз говорил о том, что нужно передавать все по константной ссылке)._

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
-	for ( int i = 0; i <= 9; ++i )
+	int bitMask = 512;
+		
+	if(0 != result && number <= bitMask)
```
ты сравниваешь указатель с `0` это может привести к неопределенным сайд эффектам. используй `nullptr` из стандарта C++0x

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
-		cleaningArray = MAX_BIN_NUMBER_LENGTH*sizeof(int);
-		memset(result, 0, cleaningArray); 
+		printf("Result = %s \n", result);
 	}
```
вот тут, к примеру, очищаешь память, после того как выходишь из цикла.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
 //-----------------------------------------------------------------------------------------
-void numberAsBinary( unsigned int number, char *_result);
+void numberAsBinary( unsigned int number, char* result);
 
```
убери пожалуйста лишьний перенос строки

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
 {
 	const int MAX_BIN_NUMBER_LENGTH = 10;
 	const int QUANTITY_DEC_DIGITS = 10;
 	unsigned int randomDecimalNamber;
```
название переменной не соответствует действительности. называется `randomDecimalNamber`, почему рандомное? Если ты его вводишь с клавиатуры? Путаешь того, кто будет читать твой код.

a-nechyporenko replied:
если ты уже решил использовать  `typedef unsigned int UInt` почему ты не применил для этой переменной

KyryloRud reviewed (commented):
Need rework.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,60 +1,46 @@
-			result[i] = '1';
+			if(number >= bitMask)
+			{
+				*(result++) = (number & bitMask) > 0 ? '1' : '0';
```
так лучше не писать, если тебе нужно инкремировать `result`, то сделай это отдельно от присваивания.
то есть:
`*result = (number & bitMask) > 0 ? '1' : '0';`
`++result;`

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
-int main(int argc, char** argv) {
+//-----------------------------------------------------------------------------------------
+void numberAsBinary( unsigned int number, char* result);
+
```
Убери пожалуйста лишьний перенос строки

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+		{
+			if(number >= bitMask)
+			{
+				*(result++) = (number & bitMask) > 0 ? '1' : '0';
```
Так не хорошо делать, вынеси инкремент отдельно:
`*result = (number & bitMask) > 0 ? '1' : '0';`
`++result;`

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
 
-int main(int argc, char** argv) {
+//-----------------------------------------------------------------------------------------
+void numberAsBinary( unsigned int number, char* result);
```
`number` в функции ты менять не будешь, по этому `const unsigned int& number` _(уже не 1 раз говорили об этом)_

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+int main() 
+{
+	const int MAX_BIN_NUMBER_LENGTH = 10;
+	const int QUANTITY_DEC_DIGITS = 10;
```
Две одинаковые переменные и хватит по функционалу только одной, так что убери одну.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,46 @@
+		cin >> randomDecimalNamber;
+		numberAsBinary(randomDecimalNamber, result);
+		printf("Result = %s \n", result);
+	}
```
тут к примеру можно очистить память из под `result`

Commit:
Fixed last bugs

KirillHelm commented on src/Task 1/main.cpp:
```diff
@@ -1,46 +1,47 @@
 {
 	const int MAX_BIN_NUMBER_LENGTH = 10;
-	const int QUANTITY_DEC_DIGITS = 10;
 	unsigned int randomDecimalNamber;
```
исправить название, уже писали в предыдущем коммите.
### Перед коммитом и пушем, пиши под каждым комметном 'Done' если ты сделал и проверь потом, исправил ли ты все что тебе писали. Если не согласен - то коменть в обсуждении.
**Иначе просто ты уже делаешь не первый коммит с одиними и теми же ошибками. Писать каждый раз надоедает. Обычно такой PR отреджектят и скажут тебе - доделывай сам.**

KirillHelm commented on src/Task 1/main.cpp:
```diff
@@ -1,46 +1,47 @@
+void numberAsBinary( const unsigned int number, char* result);
 //-----------------------------------------------------------------------------------------
+typedef unsigned int UInt;
+static const UInt MAX_NUMBER = 512u;
```
зачем создавать статическую константную переменную, если ты её пользуешь только в функции одной? Она тебе больше нигде не нужна. Либо заюзай её, либо верни в функцию.

KirillHelm commented on src/Task 1/main.cpp:
```diff
@@ -1,46 +1,47 @@
 
 	char* result;
-	result = new char[MAX_BIN_NUMBER_LENGTH];
+	result = new char[sizeof( unsigned int) * CHAR_BIT];
```
`unsigned int` и просто `int` имеют одинаковый размер. У тебя было раньше запись лучше. Ты выделял массиву размер в 10 байт. теперь же ты выделаешь: `sizeof(int) = 4;` так ещё и умножаешь его на `CHAR_BIT = 8` и в итоге ты создаешь строку равную 4*8=32 символа, для чего? Почему именно 32? если на то пошло то пиши так:
`const int stringSizePotomuchto = sizeof(int) * CHAR_BIT;`
Но тем не менее не понятно откуда взят именно такой размер...

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+
+	for(int j = 0; j < MAX_BIN_NUMBER_LENGTH; ++j)
+	{
+		cin >> randomDecimalNamber;
```
Сережа говорил не использовать `iostram`, вместо этого используй [cstdio](http://www.cplusplus.com/reference/cstdio/)

NikolayPV replied:
Сделал

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+void numberAsBinary( const unsigned int number, char* result);
+//-----------------------------------------------------------------------------------------
+typedef unsigned int UInt;
+static const UInt MAX_NUMBER = 512u;
```
зачем создавать статическую константную переменную, если ты её пользуешь только в функции одной? Она тебе больше нигде не нужна. Либо заюзай её, либо верни в функцию.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+int main() 
+{
+	const int MAX_BIN_NUMBER_LENGTH = 10;
+	unsigned int randomDecimalNamber;
```
исправить название, уже писали в предыдущем коммите.
### Перед коммитом и пушем, пиши под каждым комметном 'Done' если ты сделал и проверь потом, исправил ли ты все что тебе писали. Если не согласен - то коменть в обсуждении.
Иначе просто ты уже делаешь не первый коммит с одиними и теми же ошибками. Писать каждый раз надоедает. Обычно такой PR отреджектят и скажут тебе - доделывай сам.

NikolayPV replied:
Исправил

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+	unsigned int randomDecimalNamber;
+
+	char* result;
+	result = new char[sizeof( unsigned int) * CHAR_BIT];
```
`unsigned int` и просто `int` имеют одинаковый размер. У тебя было раньше запись лучше. Ты выделял массиву размер в 10 байт. теперь же ты выделаешь: `sizeof(int) = 4;` так ещё и умножаешь его на `CHAR_BIT = 8;` и в итоге ты создаешь строку равную 4*8=32 символа, для чего? Почему именно 32? если на то пошло то пиши так:
`const int stringSizePotomuchto = sizeof(int) * CHAR_BIT;`
Но тем не менее не понятно откуда взят именно такой размер...

KyryloRud replied:
Лучше используй `MAX_BIN_NUMBER_LENGTH`

NikolayPV replied:
это длина не двоичного числа а десятиричного

a-nechyporenko replied:
то-есть 32 чара(символа) может быть в результате?

KyryloRud replied:
Ты вводишь десятичное число в переменную `randomDecimalNamber`, а двоичное в `result`. Почему тогда ты говоришь что это длина десятичного, если в `result` будет записано двоичное?

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
 }
+
+//-----------------------------------------------------------------------------------------
+void numberAsBinary( const unsigned int number, char* _result)
```
Убери нижний прочерк из названия второй переменной

NikolayPV replied:
я меняю копию, и "_" для понимания, что работаю не с исходным result.

KyryloRud replied:
А тебе понимать не надо, за тебя это сделает проц. Ты работаешь с локальной переменной для функции.
Это так назваемая область видимости (scope).
То есть ты работать будешь внутри функции с локальной для неё result и он ничего общего с точно таким же названием переменной из main не имеет.

То есть:
```
void foo (int value)
{
   value = 15;
}
int main()
{
   int value = 22;
   foo(value);
   print(value);
   return 0;
}
```
Эта функция выведет 22, а не 15. По тому, что переменная из main будет скопирована в foo, где локальной копии для функции foo будет присвоено значение 15. После выхода из функции foo, её локальная переменная будет уничтожена. В то время как локальная переменная value для main не была изменина (её только копировали в функцию foo).

KyryloRud replied:
называть переменную начиная с нижнего прочерка - нет никакой необходимости. всегда нужно предерживаться единого стиля кодирования, иначе тебе даже будет сложно читать свой код.

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+		{
+			if(number >= bitMask)
+			{
+				*(_result++) = (number & bitMask) > 0 ? '1' : '0';
```
убери инкремент в одной строке, уже писал.
`*result = (number & bitMask) > 0 ? '1' : '0';`
`++result;`

KyryloRud commented on src/Task 1/main.cpp:
```diff
@@ -1,7 +1,47 @@
+	for(int j = 0; j < MAX_BIN_NUMBER_LENGTH; ++j)
+	{
+		cin >> randomDecimalNamber;
+		numberAsBinary(randomDecimalNamber, result);
```
а куда делася вывод на экран полученной строки? Тоже потоками cstdio _(см. коммент выше)_

NikolayPV replied:
В задании стоит возврат строки, а ее вывод является лишним, и не был предусмотрен заданием.

KyryloRud replied:
Окей, будешь снимать дамп памяти что бы мы увидели что все работает корректно? :)
То что она возращается - замечательно, только вот пруфов что оно у тебя верно работает - нет.
Добавь вывод пожалуйста, через тот же cstdio.

KyryloRud reviewed (commented):
Need to rework.

KyryloRud commented:
Названия коммитов должны соответствовать изменениям. [Вот коммит](https://github.com/NikolayPV/StudentPractice/pull/3/commits/d06c7c1f0c1be0b30967d881b795b3316d01164c), коммит меседж: **Fixed last bugs** _(пофиксены последние баги)_.
Вопросы:

- Какие последние баги?
- Пропал вывод на экран, почему это не описано?
- Какие изменения ещё были произведены?

По факту только прибавился "баг" или можно сказать изменен функционал, по тому, что нету вывода на экран, который должен быть по заданию. Но об этом не написано. Это самое основное изменение в этом коммите.

Так же не нужно плодить множество коммитов без каких либо существенных изменений (если ты сделаешь все о чем писали в коментах - это уже существенные изменения). По факту в этом коммите мало что изменилось. Если хочешь по немногу пилить функционал то делаешь себе вторую DEV ветку и работаешь в ней:
```
git checkout -b Task1Dev   //бранчуешься от Task1
... some changes ...   //кодишь...
git add .   //добавляешь изменения в коммит
git commit -m "Some changes"   //комитишь, можно комитить сколько угодно раз (то есть возращаешься к  шагу 2)
git checkout Task1   //когда сделаны все необходимые изменения переключаешься на основную ветку
git merge Task1Dev --no-commits --sqash   //сливаешь все изменения из ветки dev в основную, причем коммиты не создадуться.
git add .   //проверяешь ещё раз, все ли работает и все ли соответствует коментам на PR и добавляешь в коммит
git commit -m "More changes"   //коммитишь
git push   //пушишь в origin
```
